### DISCUSSION TITLE
browser context menu appears when using paper events since version 3.6.

### DISCUSSION THREAD

**Discussion by chrismag87** [clientIO/joint#1890] [Q&A]
Before I upgraded JointJS+ to 3.6, (I was on 3.5) when the user right clicked an element or cell, the browsers' context menu would not appear. This is the code that worked:

    'element:contextmenu': (cellView: any, evt: any) => {
        // The action after right click 
        //  Opens up a mat dialog
    },`
    
 Since this has started happening after the update, I have added this line to try and prevent it from appearing:
    
     'element:contextmenu': (cellView: any, evt: any) => {
        evt.preventDefault();
        
        // The action after right click 
        //  Opens up a mat dialog
    },`
        
  But to no prevail, can anyone add any light to this? 
  
  I cannot guarantee it is rappid but the fact it happens when right clicking the elements/cells leads me to think it is to do with the upgrade.
  
  Thanks

**Comment by kumilingus** [comment]
It sounds like a bug. Does that happen on Windows?
It could be related to this [issue](https://github.com/clientIO/joint/issues/1590) fixed by this [PR](https://github.com/clientIO/joint/commit/af875a492a226a32643b9bd2c7460974f1709a64).

**Reply by chrismag87** [reply]
Thanks for the quick reply. Yes it is windows. I assume I need to report the regression in the ticketing system @ https://www.jointjs.com/support?

**Reply by kumilingus** [reply]
No needed. This is the [issue](https://github.com/clientIO/joint/issues/1891). We'll look into it soon.

**Reply by kumilingus** [reply]
We've just tested this on Windows Chrome and could not reproduce it.

The `event` default is always prevented [here](https://github.com/clientIO/joint/blob/master/src/dia/Paper.mjs#L2026). Could you please debug whether it is the case in your app too?

**Reply by chrismag87** [reply]
Thanks @kumilingus,

I have made progress after your post though unsure why the below is happening (but will look into it so may know before you reply)

![image](https://user-images.githubusercontent.com/26097643/203022703-4059a6bd-5155-4c59-841b-7590fccccf8f.png)

You're right that the context menu event has the preventDefault call still, but when I right-click the element, it's not calling that, it's calling `contextMenuTrigger` event, and the call stack never gets to the 'evt.PreventDefault' hence why the menu shows. 

It calls this code after hitting line 2045 from the screenshot above.

```
    contextmenu: function(evt, x, y) {
        CellView.prototype.contextmenu.apply(this, arguments);
        this.notify('element:contextmenu', evt, x, y);
    },
```
When clicking the right mouse button, it hits here...

![image](https://user-images.githubusercontent.com/26097643/203027754-d592e841-d0c2-4d35-ac1f-4e3c955e05a6.png)

OK -  I decided to go back to 3.5 and I can confirm the code you referenced does get hit but never does in 3.6

![image](https://user-images.githubusercontent.com/26097643/203030732-bee157de-f96d-4a07-9bda-107de44e0ebb.png)

**This is the issue**

On v3.5, the contextmenu event is called when this line of code is fired 

![image](https://user-images.githubusercontent.com/26097643/203032710-d8ba6e1d-ed9b-42ad-a5a6-0000457311de.png)

Line 2038

However, in v.3.6, it never hits it but is there in the else

![image](https://user-images.githubusercontent.com/26097643/203032892-8e705377-9c40-4455-b2a3-6cebf6e51ed4.png)

Line 2065

Looks like some code has been added but the original logic has gone missing. 

Hope this is enough information?

Thanks

Chris

**Reply by kumilingus** [reply]
Care must be taken when debugging this. If you add a debugger to both `pointerdown` and `contextmenu`, the `contextmenu` might not be visited on Window. It's because the `contextmenu` event is triggered on `mouseup` on Windows. But you have probably release your mouse button already on `mousedown` when you hit the first debugger. Isn't that the case?

**Reply by chrismag87** [reply]
No, because I only put in a breakpoint in the evt.PreventDefault line and it just never hits on 3.6 

I did the same test in 3.5, and it does hit. So what I am saying is, I am doing the exact same actions from 3.5 to 3.6 

The code has been changed as stated above, however, I am happy to change my code to reflect the changes in 3.6, but I am 100% sure the library has changed which causes it not to work.

**Reply by kumilingus** [reply]
There is something specific to your application I'm trying to understand. It does not happen in any of the [demos](https://resources.jointjs.com/demos) here, right?

Is it happening here in this [JSFiddle](https://jsfiddle.net/kumilingus/rfs9wudp/)? If not, how to change that to reproduce the issue?

Are you using the paper `guard` option?

**Reply by chrismag87** [reply]
Hi @kumilingus 

Thanks for letting me know about the guard event. I wasn't using it but when I added it in for testing purposes, now when I right-click the node, it does now hit this line

`if (this.options.preventContextMenu) evt.preventDefault();`

![image](https://user-images.githubusercontent.com/26097643/203266444-db6bd3cc-4485-4630-a839-2be99b22eecf.png)

However the dialog I need to show doesn't appear but the browser contextmenu doesn't appear either

Thanks for the JSFiddle, I will see if I can replicate it in there for you.

**Reply by chrismag87** [reply]
Hi @kumilingus,
 
Even though the fundamental code has changed from 3.5 to 3.6 to cause the issue stated in this thread, I've created a workaround to stop the contextmenu from appearing - the application takes care of the preventDefault call and not the code in the paper.mjs file anymore.
 
![image](https://user-images.githubusercontent.com/26097643/203292851-eed5cf92-7969-4a00-8405-7ae18f6e8856.png)

Though there is still a breaking change in my opinion, I'm happy to stop the thread here. 

Thanks for the help!

**Reply by kumilingus** [reply]
It looks like a breaking change. But to be sure I'd really need to know how to reproduce it.

> However, in v.3.6, it never hits it but is there in the else
![image](https://user-images.githubusercontent.com/26097643/203032892-8e705377-9c40-4455-b2a3-6cebf6e51ed4.png)

Are you saying that `evt.button` is not `2` in your case?

**Reply by chrismag87** [reply]
It is button 2, line 2061 will hit fine. 

I believe it is to do when opening the `mat dialog `

```
'element:contextmenu': (cellView: any, evt: any) => {
     console.log('no contextmenu will open)
},`
```

This hits this line `if (this.options.preventContextMenu) evt.preventDefault();`
```
'element:contextmenu': (cellView: any, evt: any) => {
    //  Open up a mat dialog
   
      dialog.open(ExampleDialog, {
      data: {
        prop: 'foo',
      },
    });
    
},`
```
This doesn't hit this line 

`if (this.options.preventContextMenu) evt.preventDefault();`

But I don't think it's to do with the mat dialog library as such, but you're right, when I click the right mouse down, the dialog will open, but when I release the button (pointer up) the context menu then appears.

**Reply by chrismag87** [reply]
I can make the jsfiddle with the mat dialog example if this doesn't give you anymore info?

**Reply by kumilingus** [reply]
Interesting. A JSFiddle would be great, I have no idea what is the `mat dialog` does internally.

**Comment by kelanwu** [comment]
I had the same issue using Angular and JointJS V3.7.4 (upgraded from V3.5.5) on **Win**.

```
paper.on('blank:contextmenu': function(evt, x, y) {
  // Open Angular Material dialog or menu
});
```

When the Angular Material dialog or menu is open, a fullscreen overlay with high z-index will be appended at the end of html `<body>` if not exists.
![image](https://github.com/clientIO/joint/assets/32084858/9d4d0712-427b-4f8e-9719-276d516826b8)

And the issue should be introduced by this [PR](https://github.com/clientIO/joint/pull/1727/files).
Press the right mouse button (`mousedown`) -> JointJS `contextmenu` event gets triggered -> open Angular Material dialog or menu -> Release the right mouse button (`mouseup`) -> `contextmenu` event of Web API fires on the overlay to bring up the default browser context menu.

Previously, it's
Press the right mouse button (`mousedown`) -> Release the right mouse button (`mouseup`) -> JointJS `contextmenu` event gets triggered -> open Angular Material dialog or menu

`mouseup` to bring up default context menu on browser on Win is following how the context menu gets displayed on the Win OS. So in my opinion, JointJS should not make its `contextmenu` event gets triggered in a different way (`mousedown` to be consistent with Mac).

**Reply by kumilingus** [reply]
If you need to listen to original `contextmenu` event you can do the following:

```js
paper.el.addEventListener('contextmenu', function(evt) {
   // not normalized contextmenu event
}, false);
```
